doc/developer/design/20230829_topk_size_hint.md

```diff
@@ -0,0 +1,426 @@
+hint in the context of a `LATERAL` join. Probably a clearer hint there would be something
+describing the expected number of join pairs per row of the outer relation in the join.
+
+2. *Hints are about configuring low-level operators:* In this line of thinking, we accept that
+hints should apply to particular low-level operators and users need to know what operator they are
+targeting with a particular hint. Instead of `EXPECTED GROUP COUNT`, we could directly specify
+something like `EXPECTED TOPK GROUP SIZE`.
+
+The proposal in this design document is a hybrid of the two categories above in that we attach
+semantics to specific SQL clauses, which are then attached to particular HIR and MIR operators.
+We find this approach more workable that hints about low-level operators, since the same SQL
+query block can include different syntactic constructs that refer to different instances of the
+same type of low-level operator.
+
+The suggestion in MaterializeInc/materialize#18883 to add an `EXPECTED GROUP COUNT` hint is a
+proposal that follows a semantic hinting philosophy, but is higher-level in than the proposal
+in this design in that it does not refer to a specific SQL clause. Given the many variations
+that SQL syntax includes, we found it tricky to define an extensive hint set about semantic
+properties that would match the many syntactic variations for top-k. This is why this proposal
+focuses on attaching hints to the SQL syntax that encodes the top-k variants.
+
+### Non-backward compatible changes
+
+The following changes were considered:
+
+1. Changing the current behavior of `EXPECTED GROUP SIZE`: For example, we could make the
+`EXPECTED GROUP SIZE` apply only to the reduction in a single query block and force users to
+specify the other proposed hints for top-k constructs.
+
+2. Changing `EXPECTED GROUP SIZE` to `AGGREGATE GROUP SIZE`: If we were willing to change
+the behavior of `EXPECTED GROUP SIZE`, we could also change its name to reflect more clearly
+that it only applies to the reduction.
+
+The two changes above introduce operational complexity. Migration procedures would need to be
+devised and implemented where we rewrite production queries to introduce the additional
+hints instead of only the `EXPECTED GROUP SIZE` in single block queries with multiple constructs.
+Additional migration procedures would need to rewrite indexed / materialized view definitions
+in the catalog to change `EXPECTED GROUP SIZE` to `AGGREGATE GROUP SIZE`.
+
+Given that the issue is one of better UX on the specific cases where hints need to be provided
+and only in the cases where there is ambiguity, the trade-off between operational complexity
+and risk with the gain from the change leans on the direction of maintaining backwards
+compatibility, as advocated by the proposal in this design document.
+
+## Open questions
+
+* Are we OK with maintaining backwards compatibility?
```

Comment:
I think we have to. There are definitely production users today that use `EXPECTED GROUP SIZE` in their dbt models and we can't break those.

Reply:
Cool, agreed with keeping the `EXPECTED GROUP SIZE` there for backwards compatibility and changing the documentation to reference the new hints.